File: src/main/java/io/github/zenwave360/zdl/ZdlParser.java
package io.github.zenwave360.zdl;

import io.github.zenwave360.zdl.antlr.ZdlLexer;
import io.github.zenwave360.zdl.antlr.ZdlListenerImpl;
import io.github.zenwave360.zdl.antlr.ZdlModel;
import io.github.zenwave360.zdl.antlr.ZdlModelPostProcessor;
import io.github.zenwave360.zdl.antlr.ZdlModelValidator;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.ParseTreeWalker;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class ZdlParser {

    public static final List<String> STANDARD_FIELD_TYPES = List.of("String", "Integer", "Long", "int", "long", "BigDecimal", "Float", "float", "Double", "double", "Enum", "Boolean", "boolean", "LocalDate", "ZonedDateTime", "Instant", "Duration", "UUID", "byte", "byte[]", "Blob", "AnyBlob", "ImageBlob", "TextBlob");

    public static final List<String> STANDARD_VALIDATIONS = List.of("required", "unique", "min", "max", "minlength", "maxlength", "pattern");

    private List<String> standardFieldTypes = STANDARD_FIELD_TYPES;
    private List<String> extraFieldTypes = List.of();
    public ZdlParser withStandardFieldTypes(List<String> standardFieldTypes) {
        this.standardFieldTypes = standardFieldTypes;
        return this;
    }
    public ZdlParser withExtraFieldTypes(List<String> extraFieldTypes) {
        this.extraFieldTypes = extraFieldTypes;
        return this;
    }
    public ZdlModel parseModel(String model) throws IOException {
        CharStream zdl = CharStreams.fromString(model);
        ZdlLexer lexer = new ZdlLexer(zdl);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        io.github.zenwave360.zdl.antlr.ZdlParser parser = new io.github.zenwave360.zdl.antlr.ZdlParser(tokens);
        ParseTree tree = parser.zdl();
        ParseTreeWalker walker = new ParseTreeWalker();
        ZdlListenerImpl listener = new ZdlListenerImpl();
        walker.walk(listener, tree);
        var zdlModel = listener.getModel();
        zdlModel = ZdlModelPostProcessor.postProcess(zdlModel);
        try {
            zdlModel = new ZdlModelValidator()
                    .withStandardFieldTypes(standardFieldTypes)
                    .withExtraFieldTypes(extraFieldTypes)
                    .validate(zdlModel);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return zdlModel;
    }
}

File: src/commonMain/kotlin/io/zenwave360/zdl/ZdlParser.kt
package io.zenwave360.zdl

import io.zenwave360.zdl.antlr.*
import org.antlr.v4.kotlinruntime.CharStreams
import org.antlr.v4.kotlinruntime.CommonTokenStream
import org.antlr.v4.kotlinruntime.tree.ParseTreeWalker

class ZdlParser {

    companion object {
        val STANDARD_FIELD_TYPES = listOf(
            "String", "Integer", "Long", "int", "long", "BigDecimal", "Float", "float",
            "Double", "double", "Enum", "Boolean", "boolean", "Map", "LocalDate",
            "LocalDateTime", "ZonedDate", "ZonedDateTime", "Instant", "Duration", "UUID",
            "byte", "byte[]", "Blob", "AnyBlob", "ImageBlob", "TextBlob"
        )
        val STANDARD_VALIDATIONS = listOf("required", "unique", "min", "max", "minlength", "maxlength", "pattern")
    }

    private var standardFieldTypes: List<String> = STANDARD_FIELD_TYPES
    private var extraFieldTypes: List<String> = emptyList()

    fun withStandardFieldTypes(standardFieldTypes: List<String>): ZdlParser {
        this.standardFieldTypes = standardFieldTypes
        return this
    }

    fun withExtraFieldTypes(extraFieldTypes: List<String>): ZdlParser {
        this.extraFieldTypes = extraFieldTypes
        return this
    }

    fun parseModel(model: String): ZdlModel {
        val zdl = CharStreams.fromString(model)
        val lexer = ZdlLexer(zdl)
        val tokens = CommonTokenStream(lexer)
        val parser = io.zenwave360.zdl.antlr.ZdlParser(tokens)
        val listener = ZdlListenerImpl()
        val zdlRoot = parser.zdl()
        ParseTreeWalker.DEFAULT.walk(listener, zdlRoot)

        var zdlModel = listener.model
        zdlModel = ZdlModelPostProcessor.postProcess(zdlModel)
        try {
            zdlModel = ZdlModelValidator()
                .withStandardFieldTypes(standardFieldTypes)
                .withExtraFieldTypes(extraFieldTypes)
                .validate(zdlModel)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return zdlModel
    }
}

